src/hierarchies/lin_reg_uni_hierarchy.cpp

```diff
@@ -0,0 +1,192 @@
+#include "lin_reg_uni_hierarchy.hpp"
+
+#include <Eigen/Dense>
+#include <stan/math/prim/err.hpp>
+
+#include "../utils/eigen_utils.hpp"
+#include "../utils/proto_utils.hpp"
+#include "../utils/rng.hpp"
+
+void LinRegUniHierarchy::initialize() {
+  assert(prior != nullptr && "Error: prior was not provided");
+  state.regression_coeffs = hypers->mean;
+  state.var = hypers->scale / (hypers->shape + 1);
+  clear_data();
+}
+
+void LinRegUniHierarchy::clear_data() {
+  mixed_prod = Eigen::VectorXd::Zero(dim);
+  data_sum_squares = 0.0;
+  covar_sum_squares = Eigen::MatrixXd::Zero(dim, dim);
+  card = 0;
+  cluster_data_idx = std::set<int>();
+}
+
+void LinRegUniHierarchy::update_summary_statistics(
+    const Eigen::VectorXd &datum, const Eigen::VectorXd &covariate, bool add) {
+  if (add) {
+    data_sum_squares += datum(0) * datum(0);
+    covar_sum_squares += covariate * covariate.transpose();
+    mixed_prod += datum(0) * covariate;
+  } else {
+    data_sum_squares -= datum(0) * datum(0);
+    covar_sum_squares -= covariate * covariate.transpose();
+    mixed_prod -= datum(0) * covariate;
+  }
+}
+
+LinRegUniHierarchy::Hyperparams LinRegUniHierarchy::normal_invgamma_update() {
+  Hyperparams post_params;
+
+  post_params.var_scaling = covar_sum_squares + hypers->var_scaling;
+  post_params.mean = stan::math::inverse_spd(post_params.var_scaling) *
```

Comment:
This should be better done by solving the linear system
`post_params.var_scaling * x = mixed_prod + hypers->var_scaling * hypers->mean`
that can be easily done with Eigen